### ISSUE TITLE
Chinese address parsing

### ISSUE THREAD

**Issue by nathanshiaulam** [openvenues/libpostal#71]
The address parser seems to have a particularly difficult time parsing Chinese addresses. For instance:

中国，山东省，青岛市 香港东路6号，5号楼，8号室 李小方 （先生）收

provides

Result:

{
  "country": "中 国",
  "state": "山 东",
  "house": "省",
  "state": "青",
  "house": "岛 市 香 港 东",
  "road": "路",
  "house_number": "6",
  "road": "号",
  "house_number": "5",
  "house": "号",
  "road": "楼",
  "house_number": "8",
  "house": "号",
  "house_number": "室",
  "house": "李 小 方 先 生 收"
}

Has the program not been trained on Chinese addresses yet? I've added certain stopwords/other key words to resources/dictionaries/zh but it doesn't seem to have an effect.

**Comment by thatdatabaseguy** [comment]
Hey Nathan - correct, there were no Chinese addresses in the original version for two reasons:

1. The address-formatting templates that we use to format training examples didn't have formats defined for China (or Hong Kong or Taiwan or Macau) and didn't really support the notion of having more than one format per country, which is necessary for formatting CJK language addresses as the order of the components is "reversed" (depending on one's perspective) when the address is written in Chinese/Japanese/Korean and follows Western conventions when written in English or transliterated. I've since made [some](https://github.com/OpenCageData/address-formatting/pull/24) [pull requests](https://github.com/OpenCageData/address-formatting/pull/25) so we can now create training data in both formats.

2. The larger issue is that OpenStreetMap (and I think mapping in general) is [pretty much illegal in mainland China](https://en.wikipedia.org/wiki/Restrictions_on_geographic_data_in_China). We have the road network, but not a ton of addresses or venues to work with. We can probably still get to a reasonably good model of address structure using from Chinese-language addresses in Taiwan, Hong Kong and Macau, but I suspect libpostal's handling of Chinese-language addresses will be limited. Also many of the cities in China don't have polygons in OSM, which we reverse-geocode to in order to add administrative boundaries like district, prefecture, etc. via reverse-geocoding, so some addresses would just have country and not all of the admin structure.

The new model being developed in the [parser-data](https://github.com/openvenues/libpostal/tree/parser-data) branch is trained on ~61k examples in China and ~335k Chinese-language examples. For sub-building information (floor, apartment number, etc.), I've created some [per-language YAML configs](https://github.com/openvenues/libpostal/tree/parser-data/resources/addresses), a Chinese version of which could be used to randomly generate expressions like "5号楼" and "8号室". The parser model may also benefit from some Han-script-specific feature engineering since we don't do word segmentation and the parser treats each individual character as a "word" in ideographic scripts.

Incidentally, if there are any open Chinese-language address data sets that you know of (even small ones), I'd love to hear about them!

**Comment by osMatt** [comment]
> Incidentally, if there are any open Chinese-language address data sets that you know of (even small ones), I'd love to hear about them!

There are several open Taiwanese data sets at http://data.gov.tw which include address data. I am not exactly sure what kind of requirements you would have. While these are not very large, they seem to have a good variety of possible formats. For example, here are addresses of public restrooms: http://data.gov.tw/node/30794. This file includes fields for the administrative units but they do not separate house numbers from the street names. There's also a file on postal codes which features a list of street names with their respective administrative units (and postcodes): http://data.gov.tw/node/5948 (this files is also available as csv at http://download.post.gov.tw/post/download/Zip32_utf8_10501_1.csv).

> The parser model may also benefit from some Han-script-specific feature engineering

This should be very doable, as addresses are relatively well structured and a number of key words exists to separate individual components. If there is anything that I could help with, let me know.

**Comment by thatdatabaseguy** [comment]
Right on. Most of the Chinese-language address data in OSM comes from Taiwan, which has quite good coverage (probably from that same source). The one issue in Taiwanese addresses is that they tend to use traditional Chinese, whereas mainland China uses simplified, so that may not help.

There's now a [Chinese config](https://github.com/openvenues/libpostal/blob/parser-data/resources/addresses/zh.yaml) that helps generate some of the structured sub-building expressions above using both simplified and traditional Chinese in the appropriate countries. That is currently being used to build more complete training addresses for the next version of the model.

There are two additional data sets being added to the next build that should help with mainland China as well:

1. a places-only training set which will help build a more complete list of cities and their various names (Han, Pinyin, English, etc.)
2. a data set of road names and their containing polygons, so at least the common patterns in Chinese road names will be captured

**Comment by thatdatabaseguy** [comment]
Hey @nathanshiaulam, libpostal 1.0 has been released and is trained on many Chinese addresses, mostly in Taiwan and Hong Kong, but some in mainland China as well. All of the toponyms in OSM are used in the training set, so even places in China that don't have city geometries, etc. (listed as points usually) will at least be included. I've made the training sets public, (instructions for how to get it in the README), roughly 100GB of labeled sequence data for NLP. You're more than welcome to check that out and/or add other corpora to it. 

Here's the new parser's result (note that it also handles complex sub-building addresses like the one posted above):

```
> 中国，山东省，青岛市 香港东路6号，5号楼，8号室 李小方 （先生）收

Result:

{
  "country": "中国",
  "state": "山东省",
  "city": "青岛市",
  "road": "香港东路",
  "house_number": "6号",
  "level": "5号楼",
  "unit": "8号室",
  "house": "李小方 先生 收"
}
```

I've added it to our demo and test cases.

**Comment by tamade12318** [comment]
> 山东青岛市香港东路6号5号楼8号室

Result:

{
  "house": "山东青",
  "city": "岛市",
  "road": "香港东路",
  "house_number": "6号",
  "level": "5号楼",
  "unit": "8号室"
}

**Comment by albarrentine** [comment]
@tamade12318 I think the province name is missing the final "省". In [OpenStreetMap](https://www.openstreetmap.org/relation/913006) all of the Chinese names for that province are "山东省" not just "山东". If the suffix is not important in Chinese addresses (from what I've read, it seemed like it is important, but could be wrong), we can remove it for some fraction of the training data in the next release.

With toponyms (provinces, cities, etc.) it's more important that names match exactly with OSM (though OSM often contains many variations). Other components like `road`, `house`, etc. are more flexible and are predicted based on attributes of the words and label sequences.

**Comment by tamade12318** [comment]
“省” or "市" or "县" is not essential, Chinese names can ignore those suffix ,such as
________________________
山东青岛特产	山东青岛房价	山东青岛在哪里
山东青岛旅游景点	为什么要叫中国青岛	山东青岛消费水平
山东青岛师范大学	山东青岛的明星	山东青岛图片
_______________________

In written or oral Chinese expression, that is common.
How can I get Chinese OSM txt data,  and I will do something for Chinese geo NLP.
thank for your attention

**Comment by albarrentine** [comment]
Ok, we can remove the suffixes probabilistically in the next version of the training data, which would fix any parse errors related to place names without suffixes.

For raw OSM data (XML), the per-country dumps of interest are at: http://download.geofabrik.de/asia.html. Libpostal has an extensive [pre-processing pipeline](http://github.com/openvenues/libpostal/tree/master/scripts/geodata) which is necessary to produce the quality of training data that we have. Our training data is also public, and is easier to work with, although all the files are planet-wide. See https://github.com/openvenues/libpostal#training-data for instructions on downloading. Tools like `grep` can be used to select the relevant languages/countries. The three relevant corpora with training examples in mainland China are `formatted_addresses_tagged` (350k addresses in China), `formatted_places_tagged` (1.1M place names in China), and `formatted_ways_tagged` (693k geocoded street names in China).

One consideration to note is that OSM data in mainland China is limited relative to the population size because open mapping is pretty much illegal there. There's significantly more data for the Chinese language from Hong Kong and Taiwan, but they use traditional script, so may not be as useful for simplified Chinese. Much of the data we have in China are simply cities, road names, and a few addresses and points of interest that were added before the no-GPS mapping policy was put in place.
